ldap/servers/plugins/replication/windows_connection.c

```diff
@@ -776,14 +776,26 @@ send_dirsync_search(Repl_Connection *conn)
         slapi_log_err(SLAPI_LOG_REPL, windows_repl_plugin_name, "send_dirsync_search - Calling dirsync search request plugin\n");
         userfilter = windows_private_get_windows_userfilter(conn->agmt);
         if (userfilter) {
-            filter = slapi_ch_strdup(userfilter);
+            /*
+             * When we have a userfilter, we encounter an issue where a previously
+             * matching object that is *deleted* that we had synced, was not being
+             * deleted. This is because in the unfiltered case, we relied on the
+             * objectClass=*  to get everything, but when we apply a filter we are
+             * removing items that were deleted, especially if they were members of
+             * a group. As a result, we need to *always* request the isDeleted flag
+             * so that we can correct delete any remnants on our side.
+             */
+            size_t buflen = 18 + strlen(userfilter);
+            filter = slapi_ch_calloc(1, buflen);
+            snprintf(filter, buflen, "(|(isDeleted=*)%s)", userfilter);
```

Comment:
As I am not expert of Winsync, sorry for this dummy question.
This new filters looks a bit weird to me, because it will return all the matching entries (userfilter) and all the deleted entries. Is it what you expect or rather something like "(&(I(isDeleted=*)(objectclass=*))%s)".
Also are we sure that isDeleted attribute is indexed ?

Reply:
Good catch @tbordaz it looks like https://docs.microsoft.com/en-us/windows/win32/adschema/a-isdeleted is not indexed. But because of how the dirsync works, it's similar to syncrepl so it only sends us what has changed between the previous cookie and the current. 

The other possible way to do this filter is:

```
(| (userFilter ...)  (&(isDeleted=True)(userFilter ...))  )
```

That way we can lean on the userFilter to be indexed, but also means that we'll limit the isDeleted results to our userFilter set. What do you think?

Reply:
My understanding is that 'isDeleted' is similar to tombstone entry. So the initial filter would match all tombstones whether or not they match the userFilter.
IIUC the need, the second filter you proposed looks good to me
<pre>
(| (userFilter ...)  (&(isDeleted=True)(userFilter ...))  )
</pre>

Reply:
@tbordaz  See https://github.com/389ds/389-ds-base/issues/5052#issuecomment-1005388589 I think there is no way to fix this ....

Reply:
@tbordaz @Firstyear - any update on this?  Thierry do you approve this after Williams remarks?  Or do we still have concerns?

Reply:
Absolutely, Sorry I missed that it was pending on my update